packages/task/src/node/task-server.ts

```diff
@@ -0,0 +1,258 @@
+import { Task, TaskFactory, TaskProcessOptions } from './task';
+import { RawProcess, RawProcessFactory, RawProcessOptions } from '@theia/process/lib/node/raw-process';
+import { TerminalProcess, TerminalProcessFactory, TerminalProcessOptions } from '@theia/process/lib/node/terminal-process';
+import { TaskManager } from './task-manager';
+import * as fs from 'fs';
+import * as path from 'path';
+import URI from "@theia/core/lib/common/uri";
+import { FileSystem } from '@theia/filesystem/lib/common';
+import { isWindows } from '@theia/core/lib/common/os';
+import { FileUri } from "@theia/core/lib/node";
+
+@injectable()
+export class TaskServerImpl implements TaskServer {
+
+    /* Task clients to send notifications to.  */
+    protected clients: TaskClient[] = [];
+    protected taskToDispose = new Map<number, DisposableCollection>();
+    protected readonly toDispose = new DisposableCollection();
+
+    constructor(
+        @inject(ILogger) protected readonly logger: ILogger,
+        @inject(RawProcessFactory) protected readonly rawProcessFactory: RawProcessFactory,
+        @inject(TerminalProcessFactory) protected readonly terminalProcessFactory: TerminalProcessFactory,
+        @inject(TaskManager) protected readonly taskManager: TaskManager,
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(TaskFactory) protected readonly taskFactory: TaskFactory
+    ) {
+        const taskManagerListener = taskManager.onDelete(id => {
+            const toDispose = this.taskToDispose.get(id);
+            if (toDispose !== undefined) {
+                toDispose.dispose();
+                this.taskToDispose.delete(id);
+            }
+        });
+        this.toDispose.push(taskManagerListener);
+    }
+
+    dispose() {
+        // do nothing
+    }
+
+    disconnectClient(client: TaskClient) {
+        const idx = this.clients.indexOf(client);
+        if (idx > -1) {
+            this.clients.splice(idx, 1);
+        }
+    }
+
+    getRunningTasks(context?: string | undefined): Promise<TaskInfo[]> {
+        const taskinfo: TaskInfo[] = [];
+
+        const tasks = this.taskManager.getTasks(context);
+        if (tasks !== undefined) {
+            for (const task of tasks) {
+                taskinfo.push(
+                    {
+                        taskId: task.id,
+                        terminalId: (task.processType === 'terminal') ? task.process.id : undefined,
+                        processId: (task.processType === 'raw') ? task.process.id : undefined,
+                        osProcessId: task.process.pid,
+                        command: task.command,
+                        label: task.label
+                    });
+            }
+        }
+        return Promise.resolve(taskinfo);
+    }
+
+    run(task: TaskOptions, ctx?: string): Promise<TaskInfo> {
+        return this.doRun(task, ctx);
```

Comment:
why delegate to `doRun`?

Reply:
No good reason any more -> removed run(), renamed doRun() to run().